src/bcif.jl

```diff
@@ -0,0 +1,643 @@
+        if any(x -> x < type_min || x > type_max, array)
+            throw(ArgumentError("Integer values do not fit into the given dtype"))
+        end
+    end
+
+    return convert(Array{dtype}, array)
+end
+
+# Abstract encoding type
+abstract type Encoding end
+
+# ByteArrayEncoding
+mutable struct ByteArrayEncoding <: Encoding
+    type::Union{TypeCode,Nothing}
+
+    function ByteArrayEncoding(type=nothing)
+        if type !== nothing
+            type = type isa TypeCode ? type : TYPE_TO_TYPE_CODE[type]
+        end
+        new(type)
+    end
+end
+
+function encode(enc::ByteArrayEncoding, data)
+    if enc.type === nothing
+        enc.type = TYPE_TO_TYPE_CODE[eltype(data)]
+    end
+    return reinterpret(UInt8, safe_cast(data, TYPE_CODE_TO_TYPE[enc.type]))
+end
+
+function decode(enc::ByteArrayEncoding, data)
+    return reinterpret(TYPE_CODE_TO_TYPE[enc.type], data)
+end
+
+# FixedPointEncoding
+mutable struct FixedPointEncoding <: Encoding
+    factor::Float64
+    srcType::TypeCode
+
+    function FixedPointEncoding(factor; srcType=FLOAT32)
+        srcType = srcType isa TypeCode ? srcType : TYPE_TO_TYPE_CODE[srcType]
+        if !(srcType in (FLOAT32, FLOAT64))
+            throw(ArgumentError("Only floating point types are supported"))
+        end
+        new(factor, srcType)
+    end
+end
+
+function encode(enc::FixedPointEncoding, data)
+    return round.(Int32, data .* enc.factor)
+end
+
+function decode(enc::FixedPointEncoding, data)
+    return convert(Array{TYPE_CODE_TO_TYPE[enc.srcType]}, data ./ enc.factor)
+end
+
+# IntervalQuantizationEncoding
+mutable struct IntervalQuantizationEncoding <: Encoding
+    min::Float64
+    max::Float64
+    numSteps::Int
+    srcType::TypeCode
+
+    function IntervalQuantizationEncoding(min, max, numSteps; srcType=FLOAT32)
+        srcType = srcType isa TypeCode ? srcType : TYPE_TO_TYPE_CODE[srcType]
+        new(min, max, numSteps, srcType)
+    end
+end
+
+function encode(enc::IntervalQuantizationEncoding, data)
+    # Convert to normalized values between 0 and numSteps-1
+    normalized = (data .- enc.min) ./ (enc.max - enc.min) .* (enc.numSteps - 1)
+    # Clamp to valid range and convert to integers
+    indices = clamp.(round.(Int32, normalized), 0, enc.numSteps - 1)
+    return indices
+end
+
+function decode(enc::IntervalQuantizationEncoding, data)
+    # Convert indices back to values in the original range
+    normalized = data ./ Float64(enc.numSteps - 1)
+    output = normalized .* (enc.max - enc.min) .+ enc.min
+    return convert(Array{TYPE_CODE_TO_TYPE[enc.srcType]}, output)
+end
+
+# RunLengthEncoding
+mutable struct RunLengthEncoding <: Encoding
+    srcSize::Union{Int,Nothing}
+    srcType::Union{TypeCode,Nothing}
+
+    function RunLengthEncoding(; srcSize=nothing, srcType=nothing)
+        if srcType !== nothing
+            srcType = srcType isa TypeCode ? srcType : TYPE_TO_TYPE_CODE[srcType]
+        end
+        new(srcSize, srcType)
+    end
```

Comment:
Unless you're trying to disable any other way of constructing a `RunLengthEncoding` by anything other than keywords, there's probably no good reason to make this an inner constructor.